Delta-PR no.1 for multi-mode plotting

ivalaginja opened the pull request:
I suggest some changes in this PR that will make #141 slimmer while preserving all its new features.

In this PR, I:
- cleaned out and clarified the docstrings of all new functions
- simplified the new util functions that sort mode coefficients values in various ways
- removed showing of plots if not saved, otherwise there is no way to run the code silently. A `plt.show()` can always be set outside the plotting function itself if needed.
- consolidated `plot_zernike_mus()` with the already existing function `plot_segment_weights()` that does exactly that
- consolidated `plot_thermal_mus()` with the already existing function `plot_segment_weights()` that does exactly that

I also noticed that the thermal Harris mode map plotting does not work for Luvex launcher on the latest commit of #141 (9fcd41193700d53f320f4f683d5d6963dff2a663), or it is unclear how to set up the launcher for that. But we can work on this after we merge this PR into #141, and then I can do the second delta-PR, which will also implement a barebones analysis to be called from the launchers.

Note how this PR is not going into `develop` but into another branch, so there are no tests, just want your eyes on this before merging @ananyasahoo1904

ivalaginja committed:
Adapt docstring in sort_1d_mus().

ivalaginja committed:
rename sort_1d_mus()

ivalaginja committed:
Rename actuator function and remove calculable input.

ivalaginja committed:
Simplify actuator sorting function.

ivalaginja committed:
Fix up docstring of sort_1d_mus_per_actuator().

ivalaginja committed:
Clarify that these functions work for any set of mode coefficients.

ivalaginja committed:
Stay with nseg convention.

ivalaginja committed:
num_actuators not needed in plotting anymore.

ivalaginja committed:
Consolidated plot_zernike_mus() with already existing function.

ivalaginja committed:
Consolidate plot_thermal_mus() with existing function.

ivalaginja committed:
Apply same changes to LUVOIR efield launcher.

ivalaginja committed:
Clarify loop variables.

ivalaginja committed:
Rename multimode surface plotting funciton.

ivalaginja committed:
Remove unused imports and clean up imports.

asahooexo commented on pastis/launchers/run_luvex_efields.py:
```diff
@@ -1,36 +1,37 @@
 import os
-import numpy as np
 from astropy.io import fits
+import numpy as np
```
I am not sure the logic behind flipping they numpy imports.

ivalaginja replied:
I sorted them alphabetically.

asahooexo commented on pastis/launchers/run_luvex_efields.py:
```diff
@@ -1,36 +1,37 @@
-    # If Segmented Zernike Mirror, uncomment the following two lines
-    DM = 'seg_mirror' # Possible: "seg_mirror", "harris_seg_mirror", "zernike_mirror"
-    DM_SPEC = 10
+    # If using Harris deformable mirror
```
Thanks for automatizing this!

asahooexo commented on pastis/launchers/run_luvex_efields.py:
```diff
@@ -48,13 +49,21 @@
     # Calculate the static tolerances
     c_target = 6.3*1e-11
     mus = calculate_segment_constraints(pastis_matrix, c_target=c_target, coronagraph_floor=contrast_floor)
-    np.savetxt(os.path.join(dir_run, 'mus_%.2e_%d.csv' % (c_target, NUM_RINGS)), mus, delimiter=',')
+    np.savetxt(os.path.join(dir_run, f'mus_{c_target:.2e}_{NUM_RINGS:d}.csv'), mus, delimiter=',')
```
This was just a cosmetic change, matter of personal preference.

ivalaginja replied:
f-strings are easier to write, easier to read, and faster:
https://realpython.com/python-f-strings/

ananyasahoo1904 commented on pastis/launchers/run_luvoir_efields.py:
```diff
@@ -1,58 +1,71 @@
     pastis_matrix = fits.getdata(os.path.join(dir_run, 'matrix_numerical', 'pastis_matrix.fits'))
 
     # get the unaberrated coro_psf after the matrix run
     e0_psf = fits.getdata(os.path.join(dir_run, 'unaberrated_coro_psf.fits'))  # already normalized to max of direct pdf
     dh_mask = np.array(run_matrix.simulator.dh_mask.shaped)
     contrast_floor = util.dh_mean(e0_psf, dh_mask)
 
-    # set the target contrast
+    # Calculate the static tolerances
```
I will still prefer the comment "Set the target contrast", because here want to compute tolerances for a set contrast.

ananyasahoo1904 commented on pastis/launchers/run_luvex_efields.py:
```diff
@@ -48,13 +49,21 @@
 
-    ppl.plot_zernike_mus(mus, num_modes, nseg, c_target, dir_run, save=True)
-    # ppl.plot_thermal_mus(mus, num_modes, nseg, c_target, dir_run, save=True)
+    coeffs_table = util.sort_1d_mus_per_segment(mus, num_modes, nseg)
```
I feel this contradicts your statement regarding keeping launcher scripts thinner and lighter. I was expecting you to do calculation somewhere else. 
"_As for the part in the launcher, I would suggest to actually get started on a new analysis framework, rather than putting things into the launcher._" in the comment section of my PR #141

asahooexo commented on pastis/launchers/run_luvex_efields.py:
```diff
@@ -48,13 +49,21 @@
 
-    ppl.plot_zernike_mus(mus, num_modes, nseg, c_target, dir_run, save=True)
-    # ppl.plot_thermal_mus(mus, num_modes, nseg, c_target, dir_run, save=True)
+    coeffs_table = util.sort_1d_mus_per_segment(mus, num_modes, nseg)
+    mu_list = []
+    label_list = []
+    for i in range(coeffs_table.shape[0]):
+        mu_list.append(coeffs_table[i])
+        if WHICH_DM == 'seg_mirror':
+            label_list.append(f'Zernike mode {i}')
+    if WHICH_DM == 'harris_seg_mirror':
+        label_list = ['Faceplates Silvered', 'Bulk', 'Gradient Radial', 'Gradient X lateral', 'Gradient Z axial']
```
I feel this contradicts your statement regarding keeping launcher scripts thinner and lighter. The previous code plotting I added already took care of this.
"_As for the part in the launcher, I would suggest to actually get started on a new analysis framework, rather than putting things into the launcher._" in the comment section of my PR #141

ivalaginja replied:
With thinner and lighter I didn't necessarily mean a certain number of lines, but code layout. I still intend to rip this all out into an analysis module and then call it from here. This is planned for the second delta-PR, I should have clarified in the PR description here. I updated it.

Also, by adding 6 more lines here, I made an already existing function usable and made 50 lines of code with a new function redundant, reducing code length and complexity overall.

asahooexo commented on pastis/launchers/run_luvoir_efields.py:
```diff
@@ -1,58 +1,71 @@
+    if WHICH_DM == 'harris_seg_mirror':
+        label_list = ['Faceplates Silvered', 'Bulk', 'Gradient Radial', 'Gradient X lateral', 'Gradient Z axial']
 
+    ppl.plot_segment_weights(mu_list, dir_run, c_target, labels=label_list, fname=f'stat_1d_mus_{c_target:.2e}', save=True)
```
Thanks for doing this within one function, Want to make sure this function generates same plot as before. Can you please add plots generated by this function?

asahooexo replied:
Adding some plots for 3-Hex simulator with three zernike modes generated by this function.

<img width="1524" alt="Screen Shot 2022-10-26 at 2 12 28 PM" src="https://user-images.githubusercontent.com/33359812/198103897-a6e9b5f2-b895-4c6a-98dc-081090db7ed0.png">

<img width="566" alt="Screen Shot 2022-10-26 at 2 14 42 PM" src="https://user-images.githubusercontent.com/33359812/198104298-fc44215f-0dbe-4834-85ad-1e50f7b1dfad.png">


However this doesn't work for piston only aberration, I get the following message:

<img width="996" alt="Screen Shot 2022-10-26 at 2 06 33 PM" src="https://user-images.githubusercontent.com/33359812/198103543-25ffa1fb-485f-4421-b37a-16d433c97c05.png">

ivalaginja replied:
Nice catch, I'll fix this.

ivalaginja replied:
Fixed now.

asahooexo commented on pastis/util.py:
```diff
@@ -696,61 +695,72 @@ def seg_to_dm_xy(actuator_total, segment):
-    for qq in range(nmodes):
-        for kk in range(nsegments):
-            coeffs_table[qq, kk] = mus[qq + kk * nmodes]
+    coeffs_table = np.zeros([nmodes, nseg])
+    for mode in range(nmodes):
+        for seg in range(nseg):
+            coeffs_table[mode, seg] = mus[mode + seg * nmodes]
 
```
Again a change in name of the variable. I am okay with this.

ivalaginja replied:
Yes, 'qq' and 'kk' was very non-descript.

asahooexo commented on pastis/launchers/run_luvoir_efields.py:
```diff
@@ -1,58 +1,71 @@
     tel = run_matrix.simulator
-    ppl.plot_multimode_mus_surface_map(tel, mus, num_modes, tel.sm.num_actuators,
-                                       c_target, dir_run, mirror='sm', cmin=-5, cmax=5, save=True)
+    ppl.plot_multimode_surface_maps(tel, mus, num_modes, c_target, dir_run, mirror='sm', cmin=-5, cmax=5, save=True)
```
For sanity checks, Can you please add plots too? I

asahooexo replied:
I added one of the surface map plots above

asahooexo reviewed (changes requested):
@ivalaginja Thanks for taking time to review. I really like the way called "which DM" and modified  "plot_segment_weights". Over all I felt this was a of cosmetic change. My previous PR did the same thing. Can you please add some plots? I wish to make sure that these modified plotting function didn't change the way of sorting mus.

asahooexo commented:
About the commit # [9fcd411](https://github.com/spacetelescope/PASTIS/commit/9fcd41193700d53f320f4f683d5d6963dff2a663)

while using "seg_mirror" mirror, num_modes = number of localized segment aberrations = DM_SPEC.

num_modes = 10 won't work for the "'harris_seg_mirror". This is because there are only 5 thermal modes

asahooexo commented on pastis/launchers/run_luvoir_efields.py:
```diff
@@ -1,58 +1,71 @@
+
 if __name__ == '__main__':
+
+    NUM_RINGS = 1
```
NUM_RINGS is no where called in this script.

ivalaginja replied:
Ah yes, will delete.

ananyasahoo1904 commented:
@ivalaginja I generated some plots across all simulators, with different dm types. 
plot_segment_weights () is not compatible with single mode piston aberration. I have attached a screenshot of the error in the chat above. 

However, I just cross checked plot_zernike_mus() works well both for single mode piston and multi zernike modes.

asahooexo reviewed (changes requested):
@ivalaginja I generated some plots across all simulators, with different dm types.
plot_segment_weights () aberration is breaking for single mode piston only. I have attached a screenshot of the error in the chat above

Also, I just cross checked previous plot_zernike_mus() and  plot_multimode_mus_surface_maps() works well both for single mode piston and multi zernike modes.

ananyasahoo1904 commented:
@ivalaginja do you want me to merge this?

asahooexo reviewed (approved):
@ivalaginja do you want me to merge this PR?

asahooexo reviewed (changes requested):
plot_segment_weights() is breaking for piston mode only

ivalaginja committed:
Delete unused variable.

ivalaginja committed:
Make segment line plot work also for lists of length 1.

ivalaginja commented:
> @ivalaginja Thanks for taking time to review. I really like the way called "which DM" and modified "plot_segment_weights". Over all I felt this was a of cosmetic change. My previous PR did the same thing.

This PR does the same thing like #141, yes, except is uses already existing functions and removes a lot of redundant new code from #141.

> plot_segment_weights() is breaking for piston mode only

Good catch, I fixed this.

Merging now and moving on to part 2 of this.
